### DISCUSSION TITLE
Discord 機器人教學 L1

### DISCUSSION THREAD

**Discussion by ATREE01** [1101-GS4719-C/Python-Community-GS4719-C#1611] [General]
#    Discord 機器人實裝教學

##    前言
繼上次設置完機器人後，這次要教的是如何讓機器人在伺服器上有實際的功能。

#    本文
上次設定完後就可以點到Bot的這個選項，會看到有一個reset token的按鈕。
![](https://hackmd.io/_uploads/HkTfa88lp.png)

按下去之後他會產生出一個你這個機器人的token，先把他複製下來。
注意上面的提醒，這個token之後如果你忘記了，也沒辦法再找到，只能再生一個，然後之前的就會失效，這是為了避免各種安全問題的設計。

![](https://hackmd.io/_uploads/BkbcpI8gp.png)

接下來到這個頁面將這三個權限全部打開，如果想要仔細了解這些是甚麼可以參考[這篇](https://gist.github.com/advaith1/e69bcc1cdd6d0087322734451f15aa2f)文章。
![](https://hackmd.io/_uploads/rksPzPIgT.png)

接下來先創一個資料夾，裡面創一個data.json的檔案，然後把token存進去，之後會透過讀檔的方式去讀取這個token，如果你要把程式碼給別人的話才不會連token都一起給別人
![](https://hackmd.io/_uploads/BJ1h0UUla.png)

接下來在跟data.json同一個資料夾中創一個Bot.py名字你可以取你喜歡的，不過取Bot比較直觀。接下來我們會需要用叫Discord.py的套件，可以先透過下方這行指令下載，其他的套件也是同樣的下載方式。
```python
pip install discord
```

把下面這程式碼貼上執行後應該會看到這個畫面，就代表你的機器人成功上線了。

```python    
import discord
from discord.ext import commands
from discord import app_commands

from colorama import Back, Fore, Style#讓輸出變好看用的，可以不需要
import time
import json

class Bot(commands.Bot):#創建一個機器人物件，繼承了commands.Bot的屬性
    def __init__(self):
        super().__init__(command_prefix="/",intents=discord.Intents().all(),heartbeat_timeout=100)
    
    async def on_ready(self):#當機器人啟動好之後會自動執行下面這一段
        prefx = (Back.BLACK + Fore.GREEN + time.strftime("%H:%M:%S", time.localtime()) + Back.RESET + Fore.WHITE + Style.BRIGHT)
        print(prefx + ' 目前登入身份：' + Fore.BLUE + self.user.name + Fore.RESET)
        sycned = await bot.tree.sync()#這一行是之後才會用到的功能，目前沒有用。
        print(prefx + " Slash CMDS Sycned " + Fore.YELLOW + str(len(sycned)) + " Commands" + Fore.RESET)

bot = Bot()
with open("./data.json", "r", encoding='utf-8') as f:#透過讀檔的方式讀取token這樣比較安全，不會容易外洩token。
    data = json.load(f)
bot.run(data['token'])#啟動機器人。
```

![](https://hackmd.io/_uploads/SJtNQD8lp.png)

在你的伺服器中應該會看到機器人呈現上線狀態。
![](https://hackmd.io/_uploads/BJpwXP8ea.png)

接下來把這端程式碼貼上，這樣就成功定一了你的機器人的第一條指令。
名詞解釋:
interaction:是當使用者發出一個指令後會記錄一些使用者的資訊傳給機器人，像是是在哪個文字/語音頻道發出的指令。
async、await 是比較複雜的觀念，這邊先不理解也沒關係。想理解的話我這邊先放一篇[文章](https://medium.com/%E9%AB%92%E6%A1%B6%E5%AD%90/aysnc-await-%E6%95%99%E5%AD%B8%E7%AD%86%E8%A8%98-debabdb9db0e)是我隨便找的，所以不保證看得懂。

```python
bot = Bot()#貼在這行的下方。
@bot.tree.command(name='hello', description="Hello World")
async def help(interaction: discord.Interaction):
    await interaction.response.send_message("Hello Wolrd")
```

打上後從重新執行一次機器人的檔案，看到下方寫**Synced 1 commands**就代表成功加入指令了。
![](https://hackmd.io/_uploads/r1EgPPUg6.png)

接下來只要在伺服器聊天室打斜線，就會出現這個選單，可以選擇你要執行的指令了。

![](https://hackmd.io/_uploads/SJ3IDw8ep.png)

![](https://hackmd.io/_uploads/SyYSDDUea.png)

送出後就可以得到機器人回應給你的Hello World了。

![](https://hackmd.io/_uploads/BJjDDPLla.png)

這篇文章就先到此，下次再來教可以其他運用機器人的方法。

**Comment by mrweixd** [comment]
電...想知道播歌機器人的指令碼，期待下次的介紹

**Reply by ATREE01** [reply]
自己做一個播歌機器人其實蠻麻煩的，而且如果沒有常維護還會出一些奇怪的bug。一般情況的話還是建議用別人提供的就好，之後如果時間夠的話我再考慮發一下關於音樂機器人的。

**Comment by davidteng00000** [comment]
作為一個DC的重度使用者，我認為機器人真的扮演著不可或缺的角色。但是目前的機器人功能大多僅限於分發身分組...等很基礎的事情，期待未來可以加入AI元素，讓機器人變得更有趣

**Reply by lazywulf** [reply]
其實是可以的喔。模型自己建起來，機器人本地跑。
DC API 算很簡單懂的。另外，如果不想使用 Python 也可以用 JS 喔。

**Reply by ATREE01** [reply]
基本上有辦法用python做出來的事情機器人都可以實現，機器人單純是一個互動的媒介而已。
